### PR TITLE
Modernize `regrid_time` and allow setting a common calendar for decadal, yearly, and monthly data

### LINKED ISSUE #2176: regrid_time does not have promised functionality
The stated goal of `regrid_time` is 'Align time axis for cubes so they can be subtracted', but it does not accomplish this goal. It takes two inputs: a `cube`, and a `frequency`. Since it does not accept two cubes, I imagine it's meant to simply remove evidence of time information that could conflict between cubes, specified by `frequency`.

I would like to subtract the value at one month from another. To accomplish this, I call `extract_month` using two different months. Then, I understand I should apply `regrid_time` to each cube, and I imagine the correct choice of frequency is `yr`, though it's not perfectly clear from the documentation. When I follow this procedure, I am still unable to subtract one from the other, getting the error `Coordinate 'month_number' has different points for the LHS cube 'surface_temperature' and RHS cube 'surface_temperature'`.

Examining one cube at a time, I see that using `extract_month` creates an Auxiliary coordinate called `month_number`. No matter what `frequency` I use, `regrid_time` adds additional coordinates `day_of_month` and `day_of_year` rather than removing auxiliary coordinate `month_number`. 

@valeriupredoi? :upside_down_face:

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR modernizes `regrid_time` (proper documentation, type hints, proper error raising, etc.) and allows to set a custom fixed calendar for decadal, yearly, and monthly data. In addition, it makes `get_time_bounds` more flexible (allow time units that are not `days since ...`, allow all *n*-hourly data where *n* is divisor of 24).

Example:

```yaml
  regrid_time:
    calendar: standard
    units: days since 2000-01-01  # this is optional
```

Closes #2176
Related to #2106

Link to documentation: https://esmvaltool--2311.org.readthedocs.build/projects/ESMValCore/en/2311/recipe/preprocessor.html#regrid-time

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: doc/recipe/preprocessor.rst
Comment (valeriupredoi) [inline doc/recipe/preprocessor.rst:1251]:
```suggestion
* regrid_time_: Aligns the time coordinate of each dataset, against a standardized time axis.
```

**Review by valeriupredoi — APPROVED** [review]
looking very good, and a very useful addition to its functionality, cheers, Manu! Just a small docs suggestion from me, and a question about decadal bounds: isn't it better to have +/-10 years ie two straddling decades, rather than +/-5 years? the other shorter frequencies all have +/- 1 unit of frequency for bounds. just a thought - maybe the +/- 5 years is a norm I'm not aware of. Cheers, bud :beer:

**Comment by schlunma** [comment]
No, the other frequencies also have +/- 1/2 of the input frequency as bounds, see here:

```py
        if 'dec' in freq:
            min_bound = datetime(date.year - 5, 1, 1, 0, 0)
            max_bound = datetime(date.year + 5, 1, 1, 0, 0)
        elif 'yr' in freq:
            min_bound = datetime(date.year, 1, 1, 0, 0)
            max_bound = datetime(date.year + 1, 1, 1, 0, 0)
        elif 'mon' in freq or freq == 'mo':
            next_month, next_year = get_next_month(date.month, date.year)
            min_bound = datetime(date.year, date.month, 1, 0, 0)
            max_bound = datetime(next_year, next_month, 1, 0, 0)
        elif 'day' in freq:
            min_bound = date - timedelta(hours=12.0)
            max_bound = date + timedelta(hours=12.0)
        elif 'hr' in freq:
            (n_hours_str, _, _) = freq.partition('hr')
            if not n_hours_str:
                n_hours = 1
            else:
                n_hours = int(n_hours_str)
            if 24 % n_hours:
                raise NotImplementedError(
                    f"For `n`-hourly data, `n` must be a divisor of 24, got "
                    f"'{freq}'"
                )
            min_bound = date - timedelta(hours=n_hours / 2.0)
            max_bound = date + timedelta(hours=n_hours / 2.0)
```

E.g, for year it's 1 Jan for the current year and 1 Jan for the next year, etc. So I think decadal data should be +/-5 years as well.

**Comment by valeriupredoi** [comment]
you're totally right, Manu! My silly overstepping on time in me head :horse: Will merge this then, let me give it another once over

**Comment by valeriupredoi** [comment]
good to merge! Would you like to get a review from @axel-lauer as well? I see him listed in the reviewers' list

**Comment by schlunma** [comment]
Yes, if he's fine with that! I think you already tested that, right @axel-lauer?

**Review by axel-lauer — APPROVED** [review]
Yes, I tested this already. It works nicely and I used this to create a plot for our benchmarking paper.

**Comment by valeriupredoi** [comment]
great stuff, cheers, guys!

**Pull request merged**
